functorch/_src/aot_autograd.py

```diff
@@ -457,52 +922,85 @@ def add_dupe_args(args):
+    _num_mutated_metadata_only_inputs = len(_fw_metadata.metadata_mutation_input_info)
+    _num_mutated_inputs = _num_mutated_data_inputs + _num_mutated_metadata_only_inputs
+
+    # Note [Outputs that alias mutated inputs]
+    # Important: we expect the forward compiled fn to return original fw outs *and* updated inputs,
+    # **except for**
+    # - updated inputs where the only update is a metadata mutation (we don't return them in the autograd.function,
+    #   and instead we just directly replay the metadata mutation afterwards)
+    # - Outputs that alias inputs directly (same idea; we just regenerate the input alias directly afterwards)
+    _num_output_aliases_from_mutated_inps = len([
+        x for (i, x) in enumerate(_fw_metadata.aliased_output_info)
+        # Find outputs that alias their input, but where that input had a data mutation in the graph
+        if x is not None and _fw_metadata.mutated_input_info[x[0]] == MutationType.data
+    ])
 
-    out = flat_fn(*flat_args)
-    # Collect info on which output tensors require gradients,
-    # so we can mark them properly in the returned autograd.Function
-    _flat_outs_not_requiring_grad, _ = pytree.tree_flatten(
-        pytree.tree_map(
-            lambda x: isinstance(x, Tensor) and not x.requires_grad, out
-        )
-    )
+    if isinstance(out, (list, tuple)):
+        _num_non_aliased_outs = len(out[_num_mutated_data_inputs:])
+    else:
+        _num_non_aliased_outs = 1
+    assert len(_fw_metadata.requires_grad_out_info) == _num_mutated_data_inputs + _num_non_aliased_outs
+
+    # out here corresponds to the set of outputs that should be returned by the traced forward call.
+    # It includes outputs of the original forward, *and* any updated inputs due to input mutations.
+    # However, it does *not* include any outputs that are aliases of inputs, or any metadata-only input mutations.
     out = pytree.tree_map(
         lambda x: x.detach().contiguous() if isinstance(x, Tensor) else x,
         out,
     )
 
-    if isinstance(out, (list, tuple)):
-        _num_outs = len(out)
-    else:
-        _num_outs = 1
+    # This code only executes if we have graph inputs that alias each other, and one of those inputs
+    # gets its data mutated.
+    # When that happens, we replace the aliased inputs with a synthetic base, and in the traced forward
+    # we later generate the input views
+    deduped_flat_args_with_views_handled, _synthetic_base_info = merge_view_inputs(
+        deduped_flat_args, _fw_metadata.mutated_input_info)
+
+    joint_forward_backward = create_joint_forward_backward_functionalized(
+        lambda *args: flat_fn(*add_dupe_args(args)),
+        meta=_fw_metadata,
+        synthetic_base_info=_synthetic_base_info,
+    )
 
-    joint_inputs = (deduped_flat_args, out)
+    joint_inputs = (deduped_flat_args_with_views_handled, out)
 
     disable_amp = torch._C._is_any_autocast_enabled()
 
     if config.use_functionalize:
```

Comment:
do we want to kill this config (perhaps in a follow up PR..)?  or is the config going to live on but its new purpose is whether to run with decompositions?

Reply:
yeah just kill the config

Reply:
agreed, killing in the followup